stl/inc/xstring

```diff
@@ -4325,47 +4564,72 @@ private:
         _Mypair._Myval2._Myres = _BUF_SIZE - 1;
     }
 
-    void _Eos(const size_type _Newsize) { // set new length and null terminator
+    _CONSTEXPR20_CONTAINER void _Eos(const size_type _Newsize) { // set new length and null terminator
         _Traits::assign(_Mypair._Myval2._Myptr()[_Mypair._Myval2._Mysize = _Newsize], _Elem());
     }
 
-    void _Tidy_init() noexcept { // initialize basic_string data members
+    _CONSTEXPR20_CONTAINER void _Tidy_init() noexcept { // initialize basic_string data members
         _Mypair._Myval2._Mysize = 0;
-        _Mypair._Myval2._Myres  = _BUF_SIZE - 1;
-        // the _Traits::assign is last so the codegen doesn't think the char write can alias this
-        _Traits::assign(_Mypair._Myval2._Bx._Buf[0], _Elem());
+
+#ifdef __cpp_lib_constexpr_string
+        if (_STD is_constant_evaluated()) {
+            auto& _My_data         = _Mypair._Myval2;
+            _My_data._Myres        = _BUF_SIZE; // SSO disabled in constexpr context
+            auto& _Al              = _Getal();
+            const pointer _New_ptr = _Al.allocate(_BUF_SIZE + 1); // throws
+            _My_data._Bx._Ptr      = _New_ptr;
+
+            _Elem* const _Raw_new = _Unfancy(_New_ptr);
+            _Traits::assign(_Raw_new, _BUF_SIZE + 1, _Elem());
+        } else
+#endif // __cpp_lib_constexpr_string
+        {
+            _Mypair._Myval2._Myres = _BUF_SIZE - 1;
+            // the _Traits::assign is last so the codegen doesn't think the char write can alias this
+            _Traits::assign(_Mypair._Myval2._Bx._Buf[0], _Elem());
+        }
     }
 
-    void _Tidy_deallocate() noexcept { // initialize buffer, deallocating any storage
+    _CONSTEXPR20_CONTAINER void _Tidy_deallocate() noexcept { // initialize buffer, deallocating any storage
         _Mypair._Myval2._Orphan_all();
-        if (_Mypair._Myval2._Large_string_engaged()) {
+#ifdef __cpp_lib_constexpr_string
+        if (_STD is_constant_evaluated()) {
             const pointer _Ptr = _Mypair._Myval2._Bx._Ptr;
             auto& _Al          = _Getal();
             _Destroy_in_place(_Mypair._Myval2._Bx._Ptr);
             _Al.deallocate(_Ptr, _Mypair._Myval2._Myres + 1);
-        }
+        } else
+#endif // __cpp_lib_constexpr_string
+        {
+            if (_Mypair._Myval2._Large_string_engaged()) {

```

Comment:
Similarly, `_Large_string_engaged()` should always return `true` when we are being constant evaluated, so I think can leave out the added `is_constant_evaluated()` and just have that case covered by the `_Large_string_engaged()` branch!

Reply:
That is again a great catch! :+1: